includes/utils/class-amp-cache-pool.php

```diff
@@ -0,0 +1,170 @@
+<?php
+/**
+ * Class AMP_Cache_Pool
+ *
+ * @package AMP
+ */
+
+/**
+ * Cache pool abstraction based on WordPress transients.
+ *
+ * @since 1.4.0
+ */
+final class AMP_Cache_Pool {
+
+	/**
+	 * Default size of the cache pool.
+	 *
+	 * @var int
+	 */
+	const DEFAULT_POOL_SIZE = 1000;
+
+	/**
+	 * Pool map of cached entries.
+	 *
+	 * @var array
+	 */
+	private static $pool_maps = [];
+
+	/**
+	 * Index into the pool map.
+	 *
+	 * @var int
+	 */
+	private static $pool_indexes = [];
+
+	/**
+	 * Cache group to use.
+	 *
+	 * @var string
+	 */
+	private $group;
+
+	/**
+	 * Size of the cache pool.
+	 *
+	 * @var int
+	 */
+	private $size;
+
+	/**
+	 * Whether an object cache is available.
+	 *
+	 * @var bool
+	 */
+	private $is_object_cache_available;
+
+	/**
+	 * Instantiate an AMP_Cache_Pool object.
+	 *
+	 * @param string $group Optional. Group to use. Defaults to an empty string.
+	 * @param int    $size  Optional. Size of the cache pool.
+	 */
+	public function __construct( $group = '', $size = self::DEFAULT_POOL_SIZE ) {
+		$this->group = $group;
+		$this->size  = $size;
+
+		$this->is_object_cache_available = wp_using_ext_object_cache();
```

Comment:
If using an external object cache, then essentially the cache pool is not used, is that right? This assumes that that the persistent object cache has implemented some algorithm for garbage collection. 

Aside: Shouldn't WordPress transient system implement some cache replacement policy? If there was such a cache pool in core like you've implemented here, then there wouldn't be a need for this class, right?

Reply:
Yes, the manual pooling/rotation is only used with transients. Proper object caches have built-in mechanisms for this.

Reply:
I think WordPress should use something like this with a separate table, transients as they are now are a hack with several performance issues.